src/utility/retractions.jl

```diff
@@ -0,0 +1,59 @@
+#=
+Utilities for preserving the norm of (VectorInterface-compliant) vectors during optimization.
+=#
+
+"""
+    vector_retract(A, η, α)
+
+Performs a norm-preserving retraction of vector `A` along the direction `η` with step size
+`α`, giving a new vector `A´`,
+```math
+A' \\leftarrow \\cos \\left( α \\frac{||η||}{||A||} \\right) A + \\sin \\left( α \\frac{||η||}{||A||} \\right) ||A|| \\frac{η}{||η||},
+```
+and corresponding directional derivative `ξ`,
+```math
+ξ = \\cos \\left( α \\frac{||η||}{||A||} \\right) η - \\sin \\left( α \\frac{||η||}{||A||} \\right) ||η|| \\frac{A}{||A||},
+```
+such that ``\\langle A', ξ \\rangle = 0`` and ``||A'|| = ||A||``.
+"""
+function vector_retract(A, η, α)
+    n_A = norm(A)
+    n_η = norm(η)
+    cs = cos(α * n_η / n_A)
+    sn = sin(α * n_η / n_A)
+
+    A´ = add(A, η, sn * n_A / n_η, cs)
+    ξ = add(A, η, cs, -sn * n_η / n_A)
+
+    return A´, ξ
+end
+
+"""
+    vector_transport!(ξ, A, η, α, A′)
+
+Transports a direction `ξ` at `A` to a valid direction at `A´` corresponding to
+the norm-preserving retraction of `A` along `η` with step size `α`. In particular, starting
+from a direction `η` of the form
+```math
+ξ = \\left\\langle \\frac{η}{||η||}, ξ \\right\\rangle \\frac{η}{||η||} + Δξ
```

Comment:
Since this already using unicode, might as well go all the way:
```suggestion
ξ = ⟨ η / ‖η‖, ξ ⟩ η / ‖η‖ + Δξ
```
and similar below? `\langle`, `\rangle` and `\Vert`

Reply:
The only thing is that I much prefer the `\frac` look in a proper math environment, but agreed that it makes the docstring quite unreadable. I'll update throughout.